RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromZSRawProducer.cc

```diff
@@ -0,0 +1,300 @@
+
+  void SiStripClusterizerFromZSRaw::run(const FEDRawDataCollection& rawColl,
+                                        edmNew::DetSetVector<SiStripCluster>& output) {
+    ClusterFiller filler(rawColl, clusterizer_, doAPVEmulatorCheck_);
+
+    // loop over good det in cabling
+    for (auto const& idet : clusterizer_.allDets()) {
+      StripClusterizerAlgorithm::output_t::FastFiller record(output, idet.detId);
+
+      filler.fill(idet, record);
+
+      if (record.empty())
+        record.abort();
+
+    }  // end loop over dets
+  }
+
+  void ClusterFiller::fill(StripClusterizerAlgorithm::Det const& det,
+                           StripClusterizerAlgorithm::output_t::FastFiller& record) {
+    incReady();
+
+    auto idet = record.id();
+
+    COUT << "filling " << idet << std::endl;
+
+    if (!det.valid())
+      return;
+
+    incSet();
+
+    // Loop over apv-pairs of det
+    for (auto const conn : clusterizer.currentConnection(det)) {
+      if
+        UNLIKELY(!conn) continue;
+
+      const uint16_t fedId = conn->fedId();
+
+      // If fed id is null or connection is invalid continue
+      if
+        UNLIKELY(!fedId || !conn->isConnected()) { continue; }
+
+      // If Fed hasnt already been initialised, extract data and initialise
+      if (!buffers[fedId])
+        buffers[fedId] = fillBuffer(fedId, rawColl);
+      if (!buffers[fedId])
+        continue;
+      auto buffer = buffers[fedId].get();
+
+      buffer->setLegacyMode(false);
+
+      // check channel
+      const uint8_t fedCh = conn->fedCh();
+
+      if
+        UNLIKELY(!buffer->channelGood(fedCh, doAPVEmulatorCheck)) {
+          if (edm::isDebugEnabled()) {
+            std::ostringstream ss;
+            ss << "Problem unpacking channel " << fedCh << " on FED " << fedId;
+            edm::LogWarning("ClustersFromZSRawProducer") << ss.str();
+          }
+          continue;
+        }
+
+      // Determine APV std::pair number
+      uint16_t ipair = conn->apvPairNumber();
+
+      const sistrip::FEDReadoutMode mode = buffer->readoutMode();
+      auto const& ch = buffer->channel(fedCh);
+      if (mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE8 ||
+          mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE8_CMOVERRIDE) {
+        clusterizer.clustersFromZS(ch.data(), ch.offset() + 2, ch.length() - 2, ipair * 256, det, record);
+      } else if (mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED) {
+        clusterizer.clustersFromZS(ch.data(), ch.offset() + 7, ch.length() - 7, ipair * 256, det, record);
+      } else {
+        throw cms::Exception("MODE NOT SUPPORTED by ClusterFromZSRaw ") << mode;
+      }
+    }  // end loop over conn
+
+    incAct();
+
+    if (!record.empty())
+      incNoZ();
+
+    //    COUT << "filled " << record.size() << std::endl;
+    //    for (auto const& cl : record)
+    //      COUT << cl.firstStrip() << ',' << cl.amplitudes().size() << std::endl;
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant